File: golang/agents/home_test.go
// Agent-home conformance (SPEC §7E), H1–H7, against the shared fixture
// examples/persona-agent/fixture.json. Hermetic: a scripted openai-style
// RoundTripper (homeMock) implements the fixture's four models — zero network,
// zero cost — exactly as the add-subagents runtime tests do. The persona is
// REAL: a soul from files, a memory tool that persists to disk, and a heartbeat
// that only speaks when HEARTBEAT.md is due.
package agents

import (
	"encoding/json"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"testing"
	"time"
)

// ---- the shared fixture ----------------------------------------------------

type personaFixture struct {
	BootstrapDir  map[string]string `json:"bootstrapDir"`
	BootstrapOrder []string         `json:"bootstrapOrder"`
	MaxFileBytes  int               `json:"maxFileBytes"`
}

func loadFixture(t *testing.T) personaFixture {
	t.Helper()
	// golang/agents → repo root → examples/persona-agent/fixture.json
	p := filepath.Join("..", "..", "examples", "persona-agent", "fixture.json")
	b, err := os.ReadFile(p)
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var f personaFixture
	if err := json.Unmarshal(b, &f); err != nil {
		t.Fatalf("parse fixture: %v", err)
	}
	return f
}

// writeBootstrap materializes a bootstrap dir from a file map.
func writeBootstrap(t *testing.T, files map[string]string) string {
	t.Helper()
	dir := t.TempDir()
	for name, body := range files {
		if err := os.WriteFile(filepath.Join(dir, name), []byte(body), 0o644); err != nil {
			t.Fatal(err)
		}
	}
	return dir
}

// ---- scripted mock LLM (fixture models) ------------------------------------

// homeRecorder captures the `last` user message of each heartbeat call — the
// coalesce assertion inspects what actually reached one turn.
type homeRecorder struct {
	mu     sync.Mutex
	inputs []string
}

func (r *homeRecorder) record(s string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.inputs = append(r.inputs, s)
}

func (r *homeRecorder) saw(sub string) bool {
	r.mu.Lock()
	defer r.mu.Unlock()
	for _, s := range r.inputs {
		if strings.Contains(s, sub) {
			return true
		}
	}
	return false
}

// homeMock implements the fixture.json mockLLM scripts. Reuses the openai-style
// helpers (parseReq/textResp/callResp/toolCallPart) from mock_llm_test.go. The
// optional gate parks m-heartbeat until released (or the request context dies) so
// a turn can be held genuinely in-flight; the optional rec captures inputs.
type homeMock struct {
	gate chan struct{}
	rec  *homeRecorder
}

func (m homeMock) RoundTrip(req *http.Request) (*http.Response, error) {
	p := parseReq(req)
	if m.rec != nil {
		m.rec.record(p.last)
	}
	if m.gate != nil && p.model == "m-heartbeat" {
		select {
		case <-m.gate:
		case <-req.Context().Done():
			return nil, req.Context().Err()
		}
	}
	switch p.model {
	case "m-echo-soul": // reports which bootstrap sections it can see, in order
		var found []string
		for _, f := range BootstrapOrder {
			if strings.Contains(p.system, "## "+f) {
				found = append(found, f)
			}
		}
		return textResp("sections:[" + strings.Join(found, ",") + "]"), nil
	case "m-remember": // writes a memory then confirms
		if len(p.toolMsgs) == 0 {
			return callResp(toolCallPart("memory",
				map[string]any{"action": "add", "target": "user", "text": "Prefers dark roast"}, "w1")), nil
		}
		return textResp("saved: " + p.toolMsgs[0]), nil
	case "m-recall": // proves the next-session snapshot carries prior USER.md content
		if strings.Contains(p.system, "Prefers dark roast") {
			return textResp("I recall: dark roast"), nil
		}
		return textResp("no memory"), nil
	case "m-heartbeat": // speaks only when the HEARTBEAT.md rule is due
		if strings.Contains(p.system, "remind about the 3pm sync") && strings.Contains(p.last, "Heartbeat") {
			return textResp("Reminder: 3pm sync 🔔"), nil
		}
		return textResp(HeartbeatOK), nil
	default:
		return textResp("ok"), nil
	}
}

// armed reports how many After channels are currently registered on the virtual
// clock (the re-arm sync point for a deterministic heartbeat test).
func (c *fakeClock) armed() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return len(c.chs)
}

// waitArmed spins (bounded) until the virtual clock has a pending After channel.
func waitArmed(t *testing.T, c *fakeClock) {
	t.Helper()
	for i := 0; i < 500; i++ {
		if c.armed() > 0 {
			return
		}
		time.Sleep(2 * time.Millisecond)
	}
	t.Fatal("heartbeat never armed the virtual clock")
}

// falseP is a *bool pointing at false (memory opt-out).
func falseP() *bool { b := false; return &b }

// ---- H1: the directory IS the agent ----------------------------------------

func TestHomeH1BootstrapDiscoveryAndInjection(t *testing.T) {
	fx := loadFixture(t)
	dir := writeBootstrap(t, fx.BootstrapDir)

	soul, found := ComposeSoul(dir)

	// H1.1 — discovers only present files, in canonical order.
	if got := strings.Join(found, ","); got != "SOUL.md,USER.md,HEARTBEAT.md,MEMORY.md" {
		t.Errorf("H1.1 discovery order: got %q", got)
	}
	// H1.2 — injected as ## sections, SOUL before USER before HEARTBEAT before MEMORY.
	iS, iU := strings.Index(soul, "## SOUL.md"), strings.Index(soul, "## USER.md")
	iH, iM := strings.Index(soul, "## HEARTBEAT.md"), strings.Index(soul, "## MEMORY.md")
	if !(iS < iU && iU < iH && iH < iM) {
		t.Errorf("H1.2 section ordering wrong: SOUL=%d USER=%d HEARTBEAT=%d MEMORY=%d", iS, iU, iH, iM)
	}
	// H1.3 — the composed soul reaches the child's system prompt.
	ava := FromDir(dir, FromDirOptions{Model: "m-echo-soul"})
	r, _ := ava.Run(Options{Transport: homeMock{}}, "who are you?")
	if r.Text != "sections:[SOUL.md,USER.md,HEARTBEAT.md,MEMORY.md]" {
		t.Errorf("H1.3 soul did not reach child system prompt: %q", r.Text)
	}
}

// ---- H2: per-file byte cap -------------------------------------------------

func TestHomeH2PerFileByteCap(t *testing.T) {
	dir := writeBootstrap(t, map[string]string{"SOUL.md": strings.Repeat("x", 3*1024*1024)})
	soul, _ := ComposeSoul(dir)
	// H2.1 — a >2 MB file is truncated with a notice; the on-disk file is untouched.
	if !strings.Contains(soul, "[truncated: exceeds 2 MB bootstrap cap]") || len(soul) > 2*1024*1024+4096 {
		t.Errorf("H2.1 over-cap file not truncated (len=%d)", len(soul))
	}
	onDisk, _ := os.ReadFile(filepath.Join(dir, "SOUL.md"))
	if len(onDisk) != 3*1024*1024 {
		t.Errorf("H2.1 on-disk file was mutated (len=%d)", len(onDisk))
	}
}

// ---- H3: the memory tool persists to disk ----------------------------------

func TestHomeH3MemoryToolPersists(t *testing.T) {
	dir := writeBootstrap(t, map[string]string{"MEMORY.md": "- Likes green tea."})
	tool := MemoryTool(dir)
	read := func(name string) string {
		b, _ := os.ReadFile(filepath.Join(dir, name))
		return string(b)
	}

	// H3.1 — add appends an entry.
	if _, _ = tool.Execute(map[string]any{"action": "add", "text": "Likes hiking"}, nil); !strings.Contains(read("MEMORY.md"), "- Likes hiking") {
		t.Error("H3.1 add did not append")
	}
	// H3.2 — replace swaps a substring.
	if _, _ = tool.Execute(map[string]any{"action": "replace", "text": "green tea", "with": "oolong"}, nil); !strings.Contains(read("MEMORY.md"), "oolong") {
		t.Error("H3.2 replace did not swap")
	}
	// H3.3 — remove deletes an entry.
	if _, _ = tool.Execute(map[string]any{"action": "remove", "text": "- Likes hiking\n"}, nil); strings.Contains(read("MEMORY.md"), "hiking") {
		t.Error("H3.3 remove did not delete")
	}
	// H3.4 — replace of a missing substring is a loud isError; file unchanged.
	before := read("MEMORY.md")
	miss, _ := tool.Execute(map[string]any{"action": "replace", "text": "nonexistent", "with": "x"}, nil)
	if !miss.IsError || read("MEMORY.md") != before {
		t.Errorf("H3.4 missing substring not a loud error (isError=%v)", miss.IsError)
	}
	// H3.5 — target=user writes USER.md, not MEMORY.md.
	uw, _ := tool.Execute(map[string]any{"action": "add", "target": "user", "text": "Speaks Tamil"}, nil)
	if uw.IsError || !strings.Contains(read("USER.md"), "Speaks Tamil") {
		t.Error("H3.5 target=user did not write USER.md")
	}
}

// ---- H4: frozen-snapshot — a mid-session write never changes the live prompt

func TestHomeH4FrozenSnapshot(t *testing.T) {
	dir := writeBootstrap(t, map[string]string{"USER.md": "The user is new."})

	// A first run writes a memory via the tool.
	ava := FromDir(dir, FromDirOptions{Model: "m-remember"})
	ava.Run(Options{Transport: homeMock{}}, "note my coffee preference")

	// H4.1 — the write landed on disk.
	onDisk, _ := os.ReadFile(filepath.Join(dir, "USER.md"))
	if !strings.Contains(string(onDisk), "Prefers dark roast") {
		t.Errorf("H4.1 memory write did not persist: %q", string(onDisk))
	}
	// H4.2 — a SECOND, fresh FromDir re-reads the file → the snapshot carries it.
	ava2 := FromDir(dir, FromDirOptions{Model: "m-recall"})
	r2, _ := ava2.Run(Options{Transport: homeMock{}}, "what do you know about me?")
	if r2.Text != "I recall: dark roast" {
		t.Errorf("H4.2 next-session snapshot missing memory: %q", r2.Text)
	}
}

// ---- H5: heartbeat — silent OK, speaks only when due, ticks coalesce --------

func TestHomeH5Heartbeat(t *testing.T) {
	// H5.1 — a due beat surfaces to onBeat (via the real StartAgent + virtual clock).
	{
		dir := writeBootstrap(t, map[string]string{
			"SOUL.md":      "You are Ava.",
			"HEARTBEAT.md": "If it is time, remind about the 3pm sync.",
		})
		clk := newFakeClock()
		ava := FromDir(dir, FromDirOptions{Model: "m-heartbeat"})
		sa, err := StartAgent(ava, StartOptions{
			Options: Options{Transport: homeMock{}, Clock: clk},
			EveryMs: 60_000,
		})
		if err != nil {
			t.Fatal(err)
		}
		waitArmed(t, clk)
		clk.fire()
		<-sa.beat
		beats := sa.Beats()
		sa.Stop()
		if len(beats) < 1 || !strings.Contains(beats[0], "3pm sync") {
			t.Errorf("H5.1 due beat did not surface: %v", beats)
		}
	}

	// H5.2 — a HEARTBEAT_OK reply stays silent (no report).
	{
		dir := writeBootstrap(t, map[string]string{
			"SOUL.md":      "You are Ava.",
			"HEARTBEAT.md": "Nothing scheduled today.",
		})
		clk := newFakeClock()
		ava := FromDir(dir, FromDirOptions{Model: "m-heartbeat"})
		sa, err := StartAgent(ava, StartOptions{
			Options: Options{Transport: homeMock{}, Clock: clk},
			EveryMs: 60_000,
		})
		if err != nil {
			t.Fatal(err)
		}
		waitArmed(t, clk)
		clk.fire()
		<-sa.beat
		beats := sa.Beats()
		sa.Stop()
		if len(beats) != 0 {
			t.Errorf("H5.2 HEARTBEAT_OK beat was not silent: %v", beats)
		}
	}

	// H5.3 — ticks that arrive WHILE a beat's turn is in-flight coalesce into ONE
	// subsequent turn, not one turn per tick. The mock gate holds the first turn
	// genuinely running while three more ticks are posted; on release they drain
	// together, and the recorder proves a single turn saw "tick (x3 coalesced)".
	{
		dir := writeBootstrap(t, map[string]string{
			"SOUL.md":      "You are Ava.",
			"HEARTBEAT.md": "If it is time, remind about the 3pm sync.",
		})
		gate := make(chan struct{})
		rec := &homeRecorder{}
		mock := homeMock{gate: gate, rec: rec}
		ava := FromDir(dir, FromDirOptions{Model: "m-heartbeat"})
		rt := NewRuntime(Options{Transport: mock, Registry: ava.Registry()})
		h, err := rt.Spawn(rt.Root, ava.Name, nil)
		if err != nil {
			t.Fatal(err)
		}
		const hb = "Heartbeat. Read your HEARTBEAT.md section and follow it. If nothing needs attention, reply HEARTBEAT_OK."

		// Turn 1: one tick, then wake — the turn parks in the gated mock.
		rt.Post(h, InboxItem{From: "clock", Channel: "timer", Text: "tick"})
		rt.Wake(h, hb)
		if !awaitState(rt, h, StateRunning) {
			t.Fatal("H5.3 first turn never entered running")
		}
		// Three more intervals elapse while that turn is still in-flight.
		for i := 0; i < 3; i++ {
			rt.Post(h, InboxItem{From: "clock", Channel: "timer", Text: "tick"})
		}
		close(gate) // release turn 1 (and let turn 2 run freely)
		r1 := rt.Wait(h, 0)
		if !strings.Contains(r1.Text, "3pm sync") {
			t.Errorf("H5.3 first beat did not surface: %q", r1.Text)
		}
		// Turn 2: a single wake drains the three queued ticks as ONE coalesced turn.
		rt.Wake(h, hb)
		r2 := rt.Wait(h, 0)
		rt.Close(rt.Root, nil)
		if r2.Turns != 1 {
			t.Errorf("H5.3 queued ticks did not coalesce to one turn (turns=%d)", r2.Turns)
		}
		if !rec.saw("tick (x3 coalesced)") {
			t.Errorf("H5.3 three in-flight ticks did not coalesce into one turn's input; saw %v", rec.inputs)
		}
	}
}

// ---- H6: memory opt-out ----------------------------------------------------

func TestHomeH6MemoryOptOut(t *testing.T) {
	dir := writeBootstrap(t, map[string]string{"SOUL.md": "Read-only Ava."})
	ava := FromDir(dir, FromDirOptions{Memory: falseP()})
	for _, tl := range ava.Spec.Tools {
		if tl.Name == "memory" {
			t.Fatal("H6 memory tool present despite Memory=false")
		}
	}
}

// ---- H7: dream/consolidation is composition, not new surface ---------------

func TestHomeH7DreamIsComposition(t *testing.T) {
	// A "dream" agent is just FromDir + the memory tool + a HEARTBEAT.md that says
	// "fold notes into MEMORY.md" — StartAgent + MemoryTool composed, no new API.
	dir := writeBootstrap(t, map[string]string{
		"SOUL.md":      "Nightly consolidator.",
		"HEARTBEAT.md": "Consolidate: merge duplicate notes into MEMORY.md.",
	})
	dream := FromDir(dir, FromDirOptions{Model: "m-echo-soul"})
	has := false
	for _, tl := range dream.Spec.Tools {
		if tl.Name == "memory" {
			has = true
		}
	}
	if !has {
		t.Error("H7 dream agent should be fromDir + the memory tool (composition)")
	}
}

File: js/test/home.test.ts
/**
 * Agent home tests (SPEC.md §7E) — the 15 spike acceptance checks (H1–H7) rewritten
 * as shipped tests, driven by the shared `examples/persona-agent/fixture.json` where
 * it maps: the directory-is-the-agent bootstrap, the file-backed `memory` builtin with
 * frozen-snapshot semantics, and the virtual-clock heartbeat.
 *
 * Hermetic: zero network, the LLM is a scripted in-process `fetch` implementing the
 * fixture's mockLLM scripts. All heartbeat timing runs on a virtual clock.
 *
 * Run: npm run build && node --experimental-strip-types --test test/home.test.ts
 */
import { test } from "node:test"
import assert from "node:assert/strict"
import fs from "node:fs"
import os from "node:os"
import path from "node:path"
import { fileURLToPath } from "node:url"
import { agents, type Tool } from "../dist/index.js"

const { fromDir, memoryTool, composeSoul, startAgent, BOOTSTRAP_ORDER, HEARTBEAT_OK } = agents

const FIXTURE = JSON.parse(
  fs.readFileSync(
    path.resolve(fileURLToPath(import.meta.url), "../../../examples/persona-agent/fixture.json"),
    "utf8",
  ),
)

// ---------------------------------------------------------------------------
// Virtual clock (§7D: fixtures run on a virtual clock)
// ---------------------------------------------------------------------------

class VirtualClock implements agents.Clock {
  private t = 0
  private timers: Array<{ at: number; fn: () => void; done: boolean }> = []
  now(): number {
    return this.t
  }
  setTimeout(fn: () => void, ms: number): () => void {
    const timer = { at: this.t + ms, fn, done: false }
    this.timers.push(timer)
    return () => {
      timer.done = true
    }
  }
  advance(ms: number): void {
    const target = this.t + ms
    for (;;) {
      const due = this.timers.filter((x) => !x.done && x.at <= target).sort((a, b) => a.at - b.at)[0]
      if (!due) break
      this.t = due.at
      due.done = true
      due.fn()
    }
    this.t = target
  }
}

/** Flush promise chains queued behind the mock LLM's async hops. */
async function flush(rounds = 20): Promise<void> {
  for (let i = 0; i < rounds; i++) await new Promise<void>((r) => setImmediate(r))
}

// ---------------------------------------------------------------------------
// Scripted mock LLM — the fixture's mockLLM scripts (openai style, keyed by model)
// ---------------------------------------------------------------------------

function openaiResponse(message: Record<string, unknown>, tokens = FIXTURE.mockLLM.usagePerCall) {
  return new Response(JSON.stringify({ choices: [{ message: { role: "assistant", ...message } }], usage: tokens }), {
    status: 200,
    headers: { "content-type": "application/json" },
  })
}
const call = (name: string, args: Record<string, unknown>, id: string) => ({
  content: null,
  tool_calls: [{ id, type: "function", function: { name, arguments: JSON.stringify(args) } }],
})

/** Gate for the coalesce probe: `m-heartbeat-gated` blocks its turn until released,
 * so several beats overlap one in-flight turn. */
let gate = deferred()
function deferred<T = void>() {
  let resolve!: (v: T) => void
  const promise = new Promise<T>((res) => {
    resolve = res
  })
  return { promise, resolve }
}

const mockFetch: typeof fetch = async (_url, init) => {
  const body = JSON.parse(String(init?.body))
  const msgs: any[] = body.messages
  const toolMsgs = msgs.filter((m) => m.role === "tool")
  const sys = String(msgs.find((m) => m.role === "system")?.content ?? "")
  const last = String(msgs[msgs.length - 1]?.content ?? "")

  switch (body.model) {
    // m-echo-soul — reports which bootstrap sections it can see in the system prompt.
    case "m-echo-soul":
      return openaiResponse({ content: `sections:[${BOOTSTRAP_ORDER.filter((f) => sys.includes(`## ${f}`)).join(",")}]` })
    // m-remember — writes a user-memory entry, then confirms with the tool result.
    case "m-remember":
      if (toolMsgs.length === 0) return openaiResponse(call("memory", { action: "add", target: "user", text: "Prefers dark roast" }, "w1"))
      return openaiResponse({ content: `saved: ${toolMsgs[0].content}` })
    // m-recall — proves a fresh session's snapshot carries the persisted USER.md note.
    case "m-recall":
      return openaiResponse({ content: sys.includes("Prefers dark roast") ? "I recall: dark roast" : "no memory" })
    // m-heartbeat — speaks only when HEARTBEAT.md is due AND a heartbeat drove the turn.
    case "m-heartbeat":
      return openaiResponse({
        content: sys.includes("remind about the 3pm sync") && last.includes("Heartbeat") ? "Reminder: 3pm sync 🔔" : HEARTBEAT_OK,
      })
    // m-heartbeat-gated — a due heartbeat whose turn blocks on a gate (coalesce probe).
    case "m-heartbeat-gated":
      await gate.promise
      return openaiResponse({ content: "Reminder: 3pm sync 🔔" })
    default:
      return openaiResponse({ content: "ok" })
  }
}

const runOpts = { fetch: mockFetch, clock: new VirtualClock() }

// ---------------------------------------------------------------------------
// Bootstrap-dir helpers
// ---------------------------------------------------------------------------

function mkdir(files: Record<string, string>): string {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), "home-"))
  for (const [f, body] of Object.entries(files)) fs.writeFileSync(path.join(dir, f), body)
  return dir
}
/** A real bootstrap dir straight from the shared fixture. */
function fixtureDir(): string {
  return mkdir(FIXTURE.bootstrapDir)
}

// ===========================================================================
// H1 — the directory IS the agent: ordered bootstrap discovery + injection
// ===========================================================================

test("H1: ordered bootstrap discovery, ## injection, and the soul reaches the child prompt", async () => {
  const dir = fixtureDir()
  const { soul, found } = composeSoul(dir)

  // 1. only present files, in canonical order (matches the fixture's expectation)
  assert.deepEqual(found, FIXTURE.expect.H1_composedSoulSectionsInOrder)
  // 2. injected as ## sections, in order (SOUL before USER before HEARTBEAT before MEMORY)
  const order = FIXTURE.expect.H1_composedSoulSectionsInOrder.map((f: string) => soul.indexOf(`## ${f}`))
  assert.ok(order.every((v: number, i: number) => i === 0 || order[i - 1] < v), soul)
  // 3. the composed soul reaches the child's system prompt
  const ava = fromDir(dir, { model: "m-echo-soul" })
  const r = await ava.run("who are you?", runOpts)
  assert.equal(r.text, `sections:[${FIXTURE.expect.H1_composedSoulSectionsInOrder.join(",")}]`)
})

// ===========================================================================
// H2 — per-file 2 MB cap
// ===========================================================================

test("H2: a >2 MB bootstrap file is injected truncated with a notice, disk untouched", () => {
  const big = "x".repeat(3 * 1024 * 1024)
  const dir = mkdir({ "SOUL.md": big })
  const { soul } = composeSoul(dir)
  assert.ok(soul.includes("[truncated: exceeds 2 MB bootstrap cap]"))
  assert.ok(soul.length < 2.1 * 1024 * 1024, `soul length ${soul.length}`)
  // the on-disk file is untouched
  assert.equal(fs.readFileSync(path.join(dir, "SOUL.md"), "utf8").length, big.length)
})

// ===========================================================================
// H3 — the memory builtin: add / replace / remove persist to disk; loud miss
// ===========================================================================

test("H3: memory add/replace/remove persist; missing substring is a loud error; target routing", async () => {
  const dir = mkdir({ "MEMORY.md": "- Likes green tea." })
  const tool = memoryTool(dir)
  const read = (f: string) => fs.readFileSync(path.join(dir, f), "utf8")

  await tool.execute({ action: "add", text: "Likes hiking" })
  assert.ok(read("MEMORY.md").includes("- Likes hiking"))

  await tool.execute({ action: "replace", text: "green tea", with: "oolong" })
  assert.ok(read("MEMORY.md").includes("oolong") && !read("MEMORY.md").includes("green tea"))

  await tool.execute({ action: "remove", text: "- Likes hiking\n" })
  assert.ok(!read("MEMORY.md").includes("hiking"))

  const miss = await tool.execute({ action: "replace", text: "nonexistent", with: "x" })
  assert.ok(miss.isError && miss.output.includes("nonexistent"))

  const userWrite = await tool.execute({ action: "add", target: "user", text: "Speaks Tamil" })
  assert.ok(!userWrite.isError && read("USER.md").includes("Speaks Tamil"))
})

// ===========================================================================
// H4 — frozen-snapshot: a mid-session write hits disk, not the live prompt
// ===========================================================================

test("H4: memory write lands on disk and only the NEXT session's snapshot carries it", async () => {
  const dir = fixtureDir()
  const ava = fromDir(dir, { model: "m-remember" })
  await ava.run("note my coffee preference", runOpts)

  // the write landed on disk (target=user → USER.md)
  const { file, contains } = FIXTURE.expect.H4_writeLandsOnDisk
  assert.ok(fs.readFileSync(path.join(dir, file), "utf8").includes(contains))

  // a fresh fromDir re-reads the file → the snapshot now carries the note
  const next = FIXTURE.expect.H4_nextSessionSnapshotCarriesMemory
  const ava2 = fromDir(dir, { model: next.model })
  const r2 = await ava2.run("what do you know about me?", runOpts)
  assert.equal(r2.text, next.text)
})

// ===========================================================================
// H5 — heartbeat: silent HEARTBEAT_OK, substantive beat surfaces, ticks coalesce
// ===========================================================================

test("H5: HEARTBEAT_OK stays silent, a due heartbeat surfaces, and ticks coalesce", async () => {
  // (a) a persona with no due duty (no "3pm sync" in its soul) → HEARTBEAT_OK → nothing surfaces
  const quietDir = mkdir({ "SOUL.md": "You are Ava, calm." })
  const quietClock = new VirtualClock()
  const quiet = startAgent(fromDir(quietDir, { model: "m-heartbeat" }), { fetch: mockFetch, clock: quietClock }, { everyMs: 20 })
  for (let i = 0; i < 3; i++) {
    quietClock.advance(20)
    await flush()
  }
  await quiet.stop()
  assert.deepEqual(quiet.beats, [])

  // (b) a due HEARTBEAT.md → the reminder surfaces to onBeat (fixture bootstrap dir)
  const dir = fixtureDir()
  const clock = new VirtualClock()
  const onBeat: string[] = []
  const started = startAgent(fromDir(dir, { model: "m-heartbeat" }), { fetch: mockFetch, clock }, { everyMs: 20, onBeat: (t) => onBeat.push(t) })
  for (let i = 0; i < 3; i++) {
    clock.advance(20)
    await flush()
  }
  await started.stop()
  assert.ok(started.beats.length >= 1 && started.beats.every((b) => b.includes(FIXTURE.expect.H5_dueHeartbeatSurfaces.beatContains)), JSON.stringify(started.beats))
  assert.deepEqual(onBeat, started.beats)

  // (c) ticks coalesce: several intervals elapse while one gated turn is in flight →
  // the queued ticks drain as ONE turn, not one turn per tick.
  gate = deferred()
  const cDir = fixtureDir()
  const cClock = new VirtualClock()
  const coalesce = startAgent(fromDir(cDir, { model: "m-heartbeat-gated" }), { fetch: mockFetch, clock: cClock }, { everyMs: 20 })
  cClock.advance(20) // beat 1 — wakes, turn blocks on the gate
  await flush()
  cClock.advance(20) // beat 2 — handle running, tick coalesces in the inbox
  cClock.advance(20) // beat 3 — still running, tick coalesces
  await flush()
  const wakesWhileBlocked = coalesce.runtime.trace.filter((l) => l.includes("idle→running (wake)")).length
  assert.equal(wakesWhileBlocked, 1, `expected a single coalesced turn, got ${wakesWhileBlocked}`)
  gate.resolve()
  await flush()
  await coalesce.stop()
})

// ===========================================================================
// H6 — opt-out: memory:false → no memory tool in the toolkit view
// ===========================================================================

test("H6: memory:false omits the memory tool", () => {
  const dir = mkdir({ "SOUL.md": "Read-only Ava." })
  const ava = fromDir(dir, { memory: false })
  const names = (ava.spec.uses?.tools ?? []).map((t: Tool) => t.name)
  assert.ok(!names.includes("memory"))
})

// ===========================================================================
// H7 — recipe: dream/consolidation = fromDir + memory, no new surface
// ===========================================================================

test("H7: a dream/consolidation agent is just fromDir + the memory tool (composition)", () => {
  const dir = mkdir({
    "SOUL.md": "Nightly consolidator.",
    "HEARTBEAT.md": "Consolidate: merge duplicate notes into MEMORY.md via the memory tool.",
  })
  const dream = fromDir(dir, { model: "m-echo-soul" })
  const names = (dream.spec.uses?.tools ?? []).map((t: Tool) => t.name)
  assert.ok(names.includes("memory"))
})
